### DISCUSSION TITLE
Documentation on SBIC?

### DISCUSSION THREAD

**Discussion by bkzshabbaz** [polarfire-soc/polarfire-soc-documentation#33] [How to] [answered]
I'm trying to create a custom MPFSoC design with our own functionality in the FPGA fabric co-existing with the MSS.  Is there documentation on how to generate the Intel-Hex file for programming the eNVM client in Libero?  Is there documentation on the Intel-Hex file header that needs to be pre-pended to the binary ELF file before the MSS will boot from eNVM?  I only found out about this by poking in the boot mode 1 directory that gets created in SoftConsole when programming for boot mode 1.  Where does dummySbic.bin come from?  Is there a utility that creates it?

**Comment by hughbreslin** [comment, marked as answer]
Hey @bkzshabbaz Have you seen the SoftConsole MPFS Boot Mode Programmer documentation? Its available [here](https://mi-v-ecosystem.github.io/SoftConsole-Documentation/SoftConsole-v2021.3/using_softconsole/mpfs.html). The boot mode programmer is bundled with SoftConsole and its used to generate Libero friendly Intel Hex files and you can find the utility and its sources in [SoftConsole Install]/extras/mpfs.

**Comment by bkzshabbaz** [comment]
Thanks, @hughbreslin.  The Java source code also gives good insight on the format of the SBIC.  For some reason UG0881 only describes the first 208 bytes of the SBIC.

**Comment by hughbreslin** [comment]
Excellent I'm glad that helped, were you able to get enough information to close this issue? Would it have helped if our [boot-modes-fundamentals](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/fundamentals/boot-modes/boot-modes-fundamentals.md) document mentioned the boot mode programmer in SoftConsole?

**Comment by bkzshabbaz** [comment]
Is SoftConsole the only way to get the eNVM programmed with the application?  It definitely would've helped if there was some documentation to describe how Libero can be the last tool used in the development chain.

**Comment by hughbreslin** [comment]
You can also program the eNVM using Libero (which is actually what the boot mode programmer in SoftConsole does, using the fpgenprog tool in the Libero installation). To do this you need to open the "Configure Design Initialization Data and Memories" view and in the eNVM tab you can add additional clients to your bitstream:
![image](https://user-images.githubusercontent.com/44862881/155514054-04f8ac02-38d8-490b-a237-9d3fd44906f6.png)

To add your application to Libero as a hex file you're probably going to need to run the boot mode programmer in SoftConsole in dry run mode to create a Libero friendly hex file and then add that in. You can find / modify the arguments passed to the tool from the external tool configurations:
![image](https://user-images.githubusercontent.com/44862881/155514530-2ecb263b-96a2-4289-b24b-9a4aadefd765.png)

We do have this [document](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/software-development/polarfire-soc-software-tool-flow.md#full-flow) with an outline of the design flow but it could probably do with an update to include some more info on the boot modes. Thank you for the feedback, we'll have a look into the docs and how we could make the information clearer :)

**Comment by bkzshabbaz** [comment]
> To add your application to Libero as a hex file you're probably going to need to run the boot mode programmer in SoftConsole in dry run mode to create a Libero friendly hex file and then add that in. You can find / modify the arguments passed to the tool from the external tool configurations: 

I think it would be super helpful if what you just said above were placed here: https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/software-development/polarfire-soc-software-tool-flow.md#programming-envm

I'll leave it up to you to decide where to put it though :)

Thank you.

**Comment by hughbreslin** [comment]
Cool thank you for the feedback :)

**Comment by hughbreslin** [comment]
Hey @bkzshabbaz I'm closing this issue as it seems resolved, feel free to open it back up if you need any further clarification :)
